flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/RowTimeDeduplicateFunctionTest.java

```diff
@@ -0,0 +1,171 @@
+
+		testHarness.processElement(insertRecord("key1", 12, 300L));
+		testHarness.processElement(insertRecord("key2", 11, 301L));
+		testHarness.processElement(insertRecord("key3", 5, 299L));
+
+		// test 2: load snapshot state
+		testHarness.processWatermark(new Watermark(302));
+
+		assertor.assertOutputEqualsSorted("output wrong.", getExpectOutput(keepLastRow, 2), testHarness.getOutput());
+
+		// test 3: expire the state
+		testHarness.setStateTtlProcessingTime(minTtlTime.toMilliseconds() + 1);
+		testHarness.processElement(insertRecord("key1", 12, 400L));
+		testHarness.processElement(insertRecord("key2", 11, 401L));
+		testHarness.processWatermark(402);
+
+		// ("key1", 13, 99L) and ("key2", 11, 101L) had retired, thus output ("key1", 12, 200L),("key2", 11, 201L)
+		assertor.assertOutputEqualsSorted("output wrong.", getExpectOutput(keepLastRow, 3), testHarness.getOutput());
+		testHarness.close();
+	}
+
+	public void testRowTimeDeduplicateKeepLastRow() throws Exception {
+		final boolean keepLastRow = true;
+		RowTimeDeduplicateFunction func = new RowTimeDeduplicateFunction(
+				inputRowType,
+				minTtlTime.toMilliseconds(),
+				rowTimeIndex,
+				generateUpdateBefore,
+				generateInsert,
+				true);
+		KeyedProcessOperator<RowData, RowData, RowData> operator = new KeyedProcessOperator<>(func);
+		OneInputStreamOperatorTestHarness<RowData, RowData> testHarness = createTestHarness(operator);
+		testHarness.open();
+
+		testHarness.processElement(insertRecord("key1", 13, 99L));
+		testHarness.processElement(insertRecord("key1", 12, 100L));
+		testHarness.processElement(insertRecord("key2", 11, 101L));
+
+		// test 1: keep last row with row time
+		testHarness.processWatermark(new Watermark(102));
+		assertor.assertOutputEqualsSorted("output wrong.", getExpectOutput(keepLastRow, 1), testHarness.getOutput());
+
+		// do a snapshot, close and restore again
+		OperatorSubtaskState snapshot = testHarness.snapshot(0L, 0);
+		testHarness.close();
+
+		testHarness = createTestHarness(operator);
+		testHarness.setup();
+		testHarness.initializeState(snapshot);
+		testHarness.open();
+
+		testHarness.processElement(insertRecord("key1", 12, 300L));
+		testHarness.processElement(insertRecord("key2", 11, 301L));
+		testHarness.processElement(insertRecord("key3", 5, 299L));
+
+		// test 2: load snapshot state
+		testHarness.processWatermark(new Watermark(302));
+		assertor.assertOutputEqualsSorted("output wrong.", getExpectOutput(keepLastRow, 2), testHarness.getOutput());
+
+		// test 3: expire the state
+		testHarness.setStateTtlProcessingTime(minTtlTime.toMilliseconds() + 1);
+		testHarness.processElement(insertRecord("key1", 12, 400L));
+		testHarness.processElement(insertRecord("key2", 11, 401L));
+		testHarness.processWatermark(402);
+
+		// all state has expired, so the record ("key1", 12, 400L), ("key2", 12, 401L) will be INSERT message
+		assertor.assertOutputEqualsSorted("output wrong.", getExpectOutput(keepLastRow, 3), testHarness.getOutput());
```

Comment:
It's super hard to understand what the expect result is. 

I think if the result are different, then we shouldn't parameterize them into one test. We can follow the way of `TemporalRowTimeJoinOperatorTest` that resue the same test data in a common private method, but delcare differenet expected output in different test methods. 

Besides, I think we can't share the expected output for both minibatch and non-minibatch mode?